File: mobile/lib/utils/leaderboard_stats.dart
/// Sport-specific stat columns for leaderboard tables (web GuestLeaderboards parity).
///
/// Every key read here is one that recompute_player_season_stats actually
/// aggregates from player_game_stats. Table tennis previously showed a "Win%"
/// column reading `win_pct`, and profiles showed "Match wins" reading `mw` --
/// neither is ever written by scoring (only the retired showcase seed script
/// produced them), so both rendered 0 for every player forever.
library;

class LeaderboardStatCell {
  const LeaderboardStatCell({required this.label, required this.value, this.emphasis = false});
  final String label;
  final String value;
  final bool emphasis;
}

/// Reads a numeric stat key, tolerating string-encoded numbers from JSONB.
num statNum(Map<String, dynamic>? stats, String key) {
  final v = (stats ?? const {})[key];
  if (v is num) return v;
  return num.tryParse('$v') ?? 0;
}

/// Whole-percent string, e.g. 47%. Returns '—' when the denominator is 0 so an
/// unplayed stat line doesn't read as a real 0% performance.
String pct(num made, num attempted) {
  if (attempted <= 0) return '—';
  return '${((made / attempted) * 100).round()}%';
}

/// Winners-to-errors ratio. With no errors the ratio is undefined rather than
/// zero, so fall back to the winner count itself (which is how it reads in
/// practice: "9 winners, no errors").
String ratio(num numerator, num denominator) {
  if (denominator <= 0) return numerator > 0 ? numerator.toStringAsFixed(1) : '—';
  return (numerator / denominator).toStringAsFixed(1);
}

List<LeaderboardStatCell> playerStatCells(String sport, Map<String, dynamic>? stats, int gamesPlayed) {
  final gp = gamesPlayed;
  num n(String k) => statNum(stats, k);

  if (sport == 'basketball') {
    String avg(num total) => gp > 0 ? (total / gp).toStringAsFixed(1) : '0.0';
    return [
      LeaderboardStatCell(label: 'GP', value: '$gp'),
      LeaderboardStatCell(label: 'PPG', value: avg(n('total_points')), emphasis: true),
      LeaderboardStatCell(label: 'RPG', value: avg(n('total_rebounds'))),
      LeaderboardStatCell(label: 'APG', value: avg(n('total_assists'))),
      LeaderboardStatCell(label: 'SPG', value: avg(n('total_steals'))),
      LeaderboardStatCell(label: 'BPG', value: avg(n('total_blocks'))),
      LeaderboardStatCell(label: 'FG%', value: pct(n('fg_made'), n('fg_attempted'))),
    ];
  }
  if (sport == 'volleyball') {
    return [
      LeaderboardStatCell(label: 'GP', value: '$gp'),
      LeaderboardStatCell(label: 'PTS', value: '${n('pts_scored').toInt()}', emphasis: true),
      LeaderboardStatCell(label: 'Kills', value: '${n('kills').toInt()}'),
      LeaderboardStatCell(label: 'Aces', value: '${n('aces').toInt()}'),
      LeaderboardStatCell(label: 'Digs', value: '${n('digs').toInt()}'),
      LeaderboardStatCell(label: 'Blocks', value: '${n('blocks').toInt()}'),
      LeaderboardStatCell(label: 'Kill%', value: pct(n('kills'), n('attacks'))),
    ];
  }
  if (sport == 'table-tennis') {
    return [
      LeaderboardStatCell(label: 'GP', value: '$gp'),
      LeaderboardStatCell(label: 'PTS', value: '${n('pts_scored').toInt()}', emphasis: true),
      LeaderboardStatCell(label: 'Winners', value: '${n('winners').toInt()}'),
      LeaderboardStatCell(label: 'Aces', value: '${n('aces').toInt()}'),
      LeaderboardStatCell(label: 'Errors', value: '${n('errors').toInt()}'),
      LeaderboardStatCell(label: 'W/E', value: ratio(n('winners'), n('errors'))),
    ];
  }
  return [LeaderboardStatCell(label: 'GP', value: '$gp')];
}

File: apps/web/src/lib/leaderboardStats.ts
/**
 * Sport-specific stat columns for player leaderboard tables.
 *
 * Mirrors mobile/lib/utils/leaderboard_stats.dart — the two must stay in sync
 * so the same athlete reads identically on both platforms.
 *
 * Every key here is one recompute_player_season_stats actually aggregates from
 * player_game_stats. Table tennis used to show a single "Win%" column reading
 * `win_pct`, and profiles showed "Match wins" reading `mw`; neither is ever
 * written by scoring (only the retired showcase seed script produced them), so
 * both rendered 0 for every table tennis player, forever.
 */

export type LeaderboardStatCell = {
  label: string
  value: string
  /** Primary stat for the sport — rendered bold, first after GP. */
  emphasis?: boolean
}

type Stats = Record<string, number> | null | undefined

/** Reads a numeric stat key, tolerating string-encoded numbers from JSONB. */
export function statNum(stats: Stats, key: string): number {
  const v = stats?.[key]
  if (typeof v === 'number') return v
  const parsed = Number(v)
  return Number.isFinite(parsed) ? parsed : 0
}

/**
 * Whole-percent string, e.g. "47%". Returns an em dash when the denominator is
 * 0 so an unplayed stat line doesn't read as a real 0% performance.
 */
export function pct(made: number, attempted: number): string {
  if (attempted <= 0) return '—'
  return `${Math.round((made / attempted) * 100)}%`
}

/**
 * Winners-to-errors ratio. With no errors the ratio is undefined rather than
 * zero, so fall back to the winner count itself (which is how it reads in
 * practice: "9 winners, no errors").
 */
export function ratio(numerator: number, denominator: number): string {
  if (denominator <= 0) return numerator > 0 ? numerator.toFixed(1) : '—'
  return (numerator / denominator).toFixed(1)
}

export function playerStatCells(
  sport: string,
  stats: Stats,
  gamesPlayed: number,
): LeaderboardStatCell[] {
  const gp = gamesPlayed
  const n = (k: string) => statNum(stats, k)
  const avg = (total: number) => (gp > 0 ? (total / gp).toFixed(1) : '0.0')

  if (sport === 'basketball') {
    return [
      { label: 'GP', value: String(gp) },
      { label: 'PPG', value: avg(n('total_points')), emphasis: true },
      { label: 'RPG', value: avg(n('total_rebounds')) },
      { label: 'APG', value: avg(n('total_assists')) },
      { label: 'SPG', value: avg(n('total_steals')) },
      { label: 'BPG', value: avg(n('total_blocks')) },
      { label: 'FG%', value: pct(n('fg_made'), n('fg_attempted')) },
    ]
  }
  if (sport === 'volleyball') {
    return [
      { label: 'GP', value: String(gp) },
      { label: 'PTS', value: String(Math.round(n('pts_scored'))), emphasis: true },
      { label: 'Kills', value: String(Math.round(n('kills'))) },
      { label: 'Aces', value: String(Math.round(n('aces'))) },
      { label: 'Digs', value: String(Math.round(n('digs'))) },
      { label: 'Blocks', value: String(Math.round(n('blocks'))) },
      { label: 'Kill%', value: pct(n('kills'), n('attacks')) },
    ]
  }
  if (sport === 'table-tennis') {
    return [
      { label: 'GP', value: String(gp) },
      { label: 'PTS', value: String(Math.round(n('pts_scored'))), emphasis: true },
      { label: 'Winners', value: String(Math.round(n('winners'))) },
      { label: 'Aces', value: String(Math.round(n('aces'))) },
      { label: 'Errors', value: String(Math.round(n('errors'))) },
      { label: 'W/E', value: ratio(n('winners'), n('errors')) },
    ]
  }
  return [{ label: 'GP', value: String(gp) }]
}

/**
 * Compact per-athlete season summary (profile stat cards), as opposed to the
 * full leaderboard row above.
 */
export function seasonStatHighlights(
  sport: string,
  stats: Stats,
  gamesPlayed: number,
): LeaderboardStatCell[] {
  const n = (k: string) => statNum(stats, k)
  const gp = gamesPlayed <= 0 ? 1 : gamesPlayed

  if (sport === 'basketball') {
    return [
      { label: 'PPG', value: (n('total_points') / gp).toFixed(1) },
      { label: 'RPG', value: (n('total_rebounds') / gp).toFixed(1) },
      { label: 'APG', value: (n('total_assists') / gp).toFixed(1) },
      { label: 'FG%', value: pct(n('fg_made'), n('fg_attempted')) },
    ]
  }
  if (sport === 'volleyball') {
    return [
      { label: 'Kills', value: String(Math.round(n('kills'))) },
      { label: 'Aces', value: String(Math.round(n('aces'))) },
      { label: 'Digs', value: String(Math.round(n('digs'))) },
      { label: 'Blocks', value: String(Math.round(n('blocks'))) },
    ]
  }
  if (sport === 'table-tennis') {
    return [
      { label: 'PTS', value: String(Math.round(n('pts_scored'))) },
      { label: 'Winners', value: String(Math.round(n('winners'))) },
      { label: 'Aces', value: String(Math.round(n('aces'))) },
      { label: 'W/E', value: ratio(n('winners'), n('errors')) },
    ]
  }
  return [{ label: 'Games', value: String(gamesPlayed) }]
}
